src/Microsoft.Extensions.OData.Migration/UriSegmentTranslator.cs

```diff
@@ -0,0 +1,243 @@
+﻿using Microsoft.OData.Edm;
+using Microsoft.OData.UriParser;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Extensions.OData.Migration
+{
+    /// <summary>
+    /// UriSegmentTranslator contains logic to translate every kind of V3 path segment into
+    /// its corresponding V4 counterpart path segment.
+    /// </summary>
+    public class UriSegmentTranslator : Data.OData.Query.SemanticAst.PathSegmentTranslator<ODataPathSegment>

```

Comment:
I'll stop bugging you about this, but review all classes to see if they should be internal. Whenever we make it "public" we make it a public API in which we have to keep things backwards compatible.